src/Store.js

```diff
@@ -2,42 +2,141 @@ import {
+    this.clientStorage = clientStorage;
+    this.transporter.setEnvironment({ store: this, role: ROLE.CLIENT_STORAGE });
+    this.schema = schema;
+    this._options = options;
+    this._Item = Item;
+
+    this._isLoaded = {};
+    this._isLoaded.promise = new Promise((resolve, reject) => {
+      this._isLoaded.resolve = resolve;
+      this._isLoaded.reject = reject;
+    });
+
+    this.clientStorage.initialFetch([])
+    .then((csData) => {
+      csData.items.forEach((item) => {
+        item._transporterState = this._castItemTransporterState(item._transporterState);
+      });
+      this._createItems(csData.items, SOURCE.CLIENT_STORAGE);
+      return this.transporter.initialFetch(csData.items);
+    })
+    .then((tData) => {
+      tData.items.forEach((item) => {
+        item._transporterState = this._castItemTransporterState(item._transporterState);
+      });
+      this._createItems(tData.items, SOURCE.TRANSPORTER);
+      this._removeItems(tData.toDelete, SOURCE.TRANSPORTER);
+    })
+    .then(() => this._finishLoading());
+  }
+
+  // ///////////////////
+  // PUBLIC METHODS   //
+  // ///////////////////
+
+  create() {}
+  fetchAndCreate() {} // same as findOneOrFetch
+  fetch() {} // maybe? fetches again from the given SOURCE, defaults to transporter
+
+  find(identifiers) {
+    return this.items.filter(current => this._itemMatches(current, identifiers));
+  }
+  findOne(identifiers) {
+    return this.items.find(current => this._itemMatches(current, identifiers));
+  }
+  findOneOrFetch(key, source = SOURCE.TRANSPORTER) {
+    const keyIdentifier = this.schema.getKeyIdentifierFor(source.AS_TARGET);
+    if (!key[keyIdentifier]) {
+      throw new Error(`missing identifier ${keyIdentifier}`);
+    }
+    return this.findOne({ [keyIdentifier]: key[keyIdentifier] }) ||
+      this._createAndFetchFrom(key, source);
+  }
 
-    socket: {
-      url: 'http',
-    },
+  isLoaded() {
+    return this.loaded;
+  }
+
+  onceLoaded() {
+    return this._isLoaded.promise;
+  }
 
-    indexedDb: {
-      tableName: '123',
-    },
-  };
+  remove() {}
 
-  constructor({ Item, Transporter, Storage }) {
-    this.Item = Item;
+  // ///////////////////
+  // PRIVATE METHODS  //
+  // ///////////////////
 
-    if (Transporter) {
-      this.transporter = new Transporter(this.constructor.transporterOptions[Transporter.name]);
-    } else {
-      this.transporter = new EmptyTransporter();
+  _castItemTransporterState(rawState) {
+    for (const state in STATE) {
+      if (Object.prototype.hasOwnProperty.call(STATE, state)) {
+        if (state === rawState) {
+          return STATE[state];
+        }
+      }
     }
+    return undefined;
+  }
 
-    this.storage = Storage || new EmptyTransporter();
+  _createAndFetchFrom(values, source) {
+    const item = new this._Item({ store: this, autoSave: this._options.autoSave });
+    item.construct(values, { source });
```

Comment:
@tchock 
forgot to push the item to the store
  maybe we should create a second item array where all the items, that aren't in the first array (items that are being deleted when constructing the store, empty placeholder items, that are being fetched, etc), and return them if findOrFetched (maybe even the other finds?) doesn't find them in the main array. This way the main array contains only items worth displaying and the 'hidden' Item array prevents Items from being created twice while their action is in progress